### ISSUE TITLE
Add ability to save a solved Mimi model

### ISSUE THREAD

**Issue by FrankErrickson** [anthofflab/Mimi.jl#393]
I frequently solve a Mimi model and save some of the results as external files, only to realize later on that I either need additional results from the model for my work or want to double check the parameter settings for that particular model run .  This requires me to either (1) save all of the model settings in a .txt file so that I can re-run the Mimi model under the same parameterization (which can be cumbersome especially for computationally intensive models), or (2) save the model as a custom file type using a package such as `JLD` (which is not ideal to work with in cases where I plan to make modifications to the model).

It would be super convenient if there was some way to simply solve a Mimi model, save it, and at a future date load it using Mimi to explore additional parameter/variable settings for that particular model instance.  Currently, I just save a .txt file with a model's particular settings and re-run it if I need additional results, which is not ideal.

**Comment by rjplevin** [comment]
@FrankErrickson, can you provide a bit more detail please? 

* When you say "solve" a Mimi model, are you talking about optimization rather than simulation? 

* What model settings are you referring to? Just the values of all variables and parameters, or something else? To think about implementation, it would be helpful to enumerate everything you want to save and restore.

**Comment by lrennels** [comment]
@cledna same questions for you that @rjplevin asked above!  We're trying to get a good grasp for what it is you want to be able to access later.

**Comment by FrankErrickson** [comment]
@rjplevin Sure, happy to provide more details and continue the conversation beyond this if it'd be helpful.

- For the way I optimize Mimi models, I think it would just be a simulated model that I want to save.  I basically set up an optimization to find the policy solution, set this policy as a parameter in the Mimi model, and then run that model.  So when I say a solved model, what I'm referring to is whatever object Mimi creates once I set all of my parameters, create all of my connections, and then execute `run(m)`.

- At the end of the day I'd want to have all of the parameter and variable values that result from a executing `run(m)`.  It could just be a dictionary that contains all of the numerical results, but it'd be ideal if I could actually just load the Mimi model itself with the specific parameters I used so that I could re-run it if needed.  

     Maybe it'd be useful to describe the scenario I frequently find myself in and you would know best how to solve this with Mimi.  Let's say I create a new model, run it under some specific parameters, save some key results as .csv files, and then I let a week or two pass.  One of the following three things will almost definitely occur:
1. I need to do the same run, but with one parameter changed.
2.  I realized I want to save a couple of the other variables and make some plots with them.
3.  I realize something in the results doesn't look right, and I need to start digging into the parameters and any other variables that might affect the results to understand what's going on.

Currently the most convenient way to deal with 1-3 is to save all of my parameter settings in a separate file, re-construct the Mimi model, reset all of the parameters, and then run it again.  Basically redo my initial analysis.  I've used the `JLD.jl` package before to save the models, but it ended up being more trouble than it's worth.  It'd be great if I could just save the model itself and do something like `solved_model = load_mimi_model("old_model.MimiObject")`.  Even if it's just the model with parameter settings and connections (but the variables have not been solved for yet), I could load my model, quickly re-run it, and get whatever data I needed.

**Comment by rjplevin** [comment]
Frank, if we used the current state of the model to write out a new model definition (as a .jl file), including current parameter settings, so that it produces the same result when re-run, would that solve your problem? This would be more straightforward than saving the instantiated representation of the model.

**Comment by FrankErrickson** [comment]
If this .jl file can be run at a future date and I can recover my particular particular parameter and variable settings, that would solve most of my problems.  

One of the issues I encountered when using the JLD.jl package was that I'd have to construct the same Mimi Model file type before I could load my .jld file (otherwise it wouldn't recognize the Mimi model type as existing).  However, if the model itself had changed (e.g. the equations were different...which oftentimes happened), the package wouldn't recognize the file type and I ended up not being able to load my saved model results.  Maybe I was using the JLD.jl package wrong, but it seemed like it would only work in cases where the model was not longer being modified.  But this is just a long-winded way of saying that your solution would work for me if it did not depend on the same exact version of the Mimi model used to create the .jl file existing at a future date when I want to load that .jl file.

It also sounds like this wouldn't solve Problem 1 above (needing to do the same exact run, but with one parameter changed), but that's not so much of a big problem on my end... more wishful thinking because it's a case that frequently pops up.

**Comment by rjplevin** [comment]
Yeah, JLD is pretty low-level, and will result in unreadable files if the types evolve (which they will, for a while, I'm sure.) I think the approach of generating the source code for an equivalent model is going to be more robust (and easier to update if there's an API change, for example.)

As for changing a parameter, this would be easy once you've defined the model: just update any parameters and run the model. Or am I missing something here?

**Comment by FrankErrickson** [comment]
I should explain the changing a parameter issue in better detail.  The issue I ran into came up over the last year working with a group on a new model.  I'd do some model runs and save the results, then over the next month or two we continue working on the model and I do some more runs.  Then I usually get an email saying, "These results look great, how do they compare to the previous version of the model from a few months ago if we try them both under a variety of parameter settings."  

Since the model has been updated a lot, it's not super quick to carry out these model runs/comparisons since I don't have quick access to the old version.  I just thought if I could save an instance of the Mimi model itself as a file (that could be loaded, re-run with new parameters, etc.), it would make model comparison over time quicker/easier on my end.  Of course I can always do this through Git, it'd just be easier to load a file for me rather than sorting through my commit history.  But this seems like more of a niche issue, especially if it's non-trivial to set up on the Mimi side of things.

**Comment by lrennels** [comment]
An update from discussion 5/10/2019

The thought is to save all parameter and variable settings and results, as an intermediate solution instead of trying to save a model that could actually be rerun. This would be helpful for users such as Frank, who need to be able to pass the information around to users who use Matlab etc.

Thinking through options for file format:
- CSV - columns including name, time, region, value ... worry is that this could get too big
- feather
